### PR TITLE
Enable text for scattergl trace

### PR DESCRIPTION
**Pull request by dy**
Implements https://github.com/plotly/plotly.js/issues/2501

* [x] Text rendering via [gl-text](https://ghub.io/a-vis/text)
  * [x] antialiasing
  * [x] kerning
  * [x] 1e6 characters (test)
  * [x] horiz/vert alignment
* [x] `gl_text_chart_*` mocks

### CODE REVIEW

**Comment by etpinard** [comment]
Nice @dy :tada: 

Anything in particular you'd like me to help out with?

File: src/traces/scattergl/convert.js [hunk 1]
```diff
@@ -40,6 +41,46 @@ function convertStyle(gd, trace) {
...
+            textpos = textpos.split(/\s+/)
+            switch (textpos[1]) {
+                case 'left':
+                    textOptions.align = 'right';
+                    break;
+                case 'right':
+                    textOptions.align = 'left';
+                    break;
+                case 'center':
+                case 'centre':
```
Comment (etpinard) [inline src/traces/scattergl/convert.js]: That's not a valid value. We should implement the same `textposition` values as scatter:

```
            'top left', 'top center', 'top right',
            'middle left', 'middle center', 'middle right',
            'bottom left', 'bottom center', 'bottom right'
```

**Comment by etpinard** [comment]
@dy regl-text appears pretty slow at 3000 `text` entries.

Try:

```js
Plotly.d3.json('https://plot.ly/~jackp/18395.json', (err, fig) => {
  fig.data[1].type = 'scattergl'
  fig.config = {scrollZoom: true}
  console.time('regl-text')
  Plotly.newPlot(gd, fig)
  console.timeEnd('regl-text')
})
```

I get 5000-6000ms

File: src/traces/scattergl/attributes.js [hunk 2]
```diff
@@ -77,10 +81,12 @@ var attrs = module.exports = overrideAll({
...
 
     selected: {
-        marker: scatterAttrs.selected.marker
+        marker: scatterAttrs.selected.marker,
+        textfont: scatterAttrs.selected.textfont
     },
     unselected: {
-        marker: scatterAttrs.unselected.marker
+        marker: scatterAttrs.unselected.marker,
+        textfont: scatterAttrs.unselected.textfont
```
Comment (etpinard) [inline src/traces/scattergl/attributes.js]: Does this work? And what about `selected.textfont`?

File: src/components/drawing/index.js [hunk 3]
```diff
@@ -622,8 +624,7 @@ function textPointPosition(s, textPosition, fontSize, markerRadius) {
 
     var numLines = (svgTextUtils.lineCount(s) - 1) * LINE_SPACING + 1;
     var dx = TEXTOFFSETSIGN[h] * r;
-    var dy = fontSize * 0.75 + TEXTOFFSETSIGN[v] * r +
-            (TEXTOFFSETSIGN[v] - 1) * numLines * fontSize / 2;
+    var dy = fontSize * 0.75 + TEXTOFFSETSIGN[v] * r + (TEXTOFFSETSIGN[v] - 1) * numLines * fontSize / 2;
```
Comment (etpinard) [inline src/components/drawing/index.js]: Can you try not committing things like that in the future? Thank you.

File: src/traces/scattergl/attributes.js [hunk 4]
```diff
@@ -37,9 +37,13 @@ var attrs = module.exports = overrideAll({
...
         ].join(' ')
     }),
+
+    textposition: scatterAttrs.textposition,
+    textfont: scatterAttrs.textfont,
+
     mode: {
         valType: 'flaglist',
-        flags: ['lines', 'markers'],
+        flags: ['lines', 'markers', 'text'],
```
Comment (etpinard) [inline src/traces/scattergl/attributes.js]: Now that scattergl supports on-graph, we can add `hovertext` like in svg scatter traces:

https://github.com/plotly/plotly.js/blob/6ac47eca310d15e412cf9f6aa41652e0e33501e7/src/traces/scatter/attributes.js#L91-L104

File: src/traces/scattergl/index.js [hunk 5]
```diff
@@ -213,6 +220,11 @@ function sceneUpdate(gd, subplot) {
             if(scene.line2d) scene.line2d.update(opts);
             if(scene.error2d) scene.error2d.update(opts.concat(opts));
             if(scene.select2d) scene.select2d.update(opts);
+            if(scene.glText) {
```
Comment (etpinard) [inline src/traces/scattergl/index.js]: Can you :hocho: this out?

File: src/traces/scattergl/index.js [hunk 6]
```diff
@@ -161,6 +165,7 @@ function sceneOptions(gd, subplot, trace, positions, x, y) {
     return opts;
 }
 
+
```
Comment (etpinard) [inline src/traces/scattergl/index.js]: Again here. Please don't commit lines like this.

File: tasks/header.js [hunk 7]
```diff
@@ -87,7 +87,10 @@ function updateHeadersInSrcFiles() {
     });
 
     function isCorrect(header) {
-        return (header.value === licenseStr);
+        return (
+            header.value.replace(/\s+$/gm, '') ===
```
Comment (etpinard) [inline tasks/header.js]: Is this for windows compat?

If so, I would prefer a different solution. Trimming space here could lead to false positive assertion.

**Comment by etpinard** [comment]
TODO:

- [ ] `npm run build` is failing at the plot-schema step. Some gltext dependency does not like jsdom it seems.

File: test/image/mocks/gl_text_chart_arrays.json [hunk 8]
```diff
@@ -0,0 +1,144 @@
+{
+    "data": [
```
Comment (etpinard) [inline test/image/mocks/gl_text_chart_arrays.json]: gl2d mocks use the `gl2d_` prefix

@dy do you mind renaming these new mocks?

**Comment by etpinard** [comment]
TODO: 

- [ ] pan doesn't work.

File: tasks/header.js (hunk 7 above)
Comment (etpinard) [reply tasks/header.js]: _never mind_. Thanks for the fix!

**Comment by dy** [comment]
* [ ] firefox text
* [ ] ubuntu test

**Comment by etpinard** [comment]
- [ ] selection doesn't work

File: src/traces/scattergl/attributes.js (hunk 2 above)
Comment (etpinard) [reply src/traces/scattergl/attributes.js]: Confirmed. `selected.textfont.color` and `unselected.textfont.color` do **not** work.

File: src/traces/scattergl/index.js (hunk 6 above)
Comment (dy) [reply src/traces/scattergl/index.js]: Ok, but in this case all functions are separated by 2 new lines and here by just one, so hope you don't mind a bit of perfectionism

File: test/image/mocks/gl_text_chart_arrays.json (hunk 8 above)
Comment (dy) [reply test/image/mocks/gl_text_chart_arrays.json]: Fixed

**Comment by dy** [comment]
* [ ] optimize performance of big N of text instances

**Comment by etpinard** [comment]
@dy any progress on this since last Thursday?

**Comment by dy** [comment]
@etpinard sure. I reworked panning, so now it is as fast as regl-scatter2d. It remains to add batch rendering for multiple colors, fonts etc. and we're all set. ~2-3 hours.

**Comment by dy** [comment]
http://192.168.1.61:9967/

**Comment by dy** [comment]
* [ ] ubuntu blurry text
* [ ] text is cut in firefox
* [ ] text_chart_single-string: missing top line
* [ ] gl-text full texts failure

**Comment by etpinard** [comment]
@dy I'm getting 

image

of your latest commit when during to select on all the `gl2d_text_chart*` mocks we have. Are you planning on working on this, or just I try to fix this?

**Comment by dy** [comment]
Will fix that in the next 2h

**Comment by etpinard** [comment]
@dy I went ahead and tried to fix a few bugs I noticed in https://github.com/plotly/plotly.js/compare/textgl...textgl-etienne, mostly related to marker and text keys in `scene.selectedOptions` conflicting.

File: src/traces/scattergl/index.js [hunk 9]
```diff
@@ -807,7 +854,68 @@ function selectPoints(searchInfo, polygon) {
...
         }
         // we should turn scatter2d into unselected once we have any points selected
-        scene.scatter2d.update(scene.unselectedOptions);
+        if(hasMarkers) {
+            scene.scatter2d.update(scene.unselectedOptions);
+        }
+    }
+
+    // update texts selection
+    if(hasText) {
```
Comment (etpinard) [inline src/traces/scattergl/index.js]: We'll have to move this to out of `selectPoints` and into `scene.plot` to make set sure e.g. `Plotly.restyle(gd, 'selected.marker.textfont.color', 'red')` works.

Comment (etpinard) [reply src/traces/scattergl/index.js]: Fixed in bdc8f4a1a96ed498f0851b775e55ab974bd36265

File: src/traces/scattergl/attributes.js (hunk 2 above)
Comment (etpinard) [reply src/traces/scattergl/attributes.js]: mostly done in https://github.com/plotly/plotly.js/commit/1a640051622f47165fd9edfa9cba073e95447f50

File: src/traces/scattergl/attributes.js (hunk 4 above)
Comment (etpinard) [reply src/traces/scattergl/attributes.js]: done in https://github.com/plotly/plotly.js/commit/959a5554794586e941576e553de6b7e3cbfb74d3

**Comment by etpinard** [comment]
@dy, I'm pretty pleased with how scattergl text behaves on my branch [`textgl-etienne`](https://github.com/plotly/plotly.js/compare/textgl...textgl-etienne). Would it be ok if I merged it in this PR branch?

The only :no_entry: blocking :no_entry:  bug is selections for `mode: 'text'`. On my branch, selections are now working ok on when markers are also visible (e.g. `mode: 'markers+text'`), but fail when only text items are drawn:

peek 2018-06-29 15-52

The example is https://github.com/plotly/plotly.js/pull/2737#issuecomment-398082311 also suffers from the same problem.

-----------

I noticed :two: other bugs along the way (https://github.com/plotly/plotly.js/issues/2765 and https://github.com/plotly/plotly.js/issues/2767) which would be nice to fix and :lock: along with https://github.com/plotly/plotly.js/issues/2731 before 1.39.0, but they can wait the following patch release if you don't have the time.

**Comment by etpinard** [comment]
@dy I think I found a way to fix selections for `mode: 'text'` traces in https://github.com/plotly/plotly.js/commit/3af4d2f57c8a5c7f311a722546d500dab0e61aff

peek 2018-06-29 16-28

What do you think?

**Comment by etpinard** [comment]
(@dy note that I'll be offline until tuesday July 2nd)

**Comment by etpinard** [comment]
No news from @dy, merging in https://github.com/plotly/plotly.js/compare/textgl...textgl-etienne to make reviewing easier.

File: src/traces/scattergl/index.js [hunk 10]
```diff
@@ -132,9 +136,10 @@ function calc(gd, trace) {
     return [{x: false, y: false, t: stash, trace: trace}];
 }
 
+
 // create scene options
 function sceneOptions(gd, subplot, trace, positions, x, y) {
-    var opts = convertStyle(gd, trace);
+    var opts = convert.style(gd, trace);
```
Comment (dy) [inline src/traces/scattergl/index.js]: Nice!

**Comment by dy** [comment]
Nice test case. Thanks for optimizing the code @etpinard, looks neat. :dancer:

**Comment by etpinard** [comment]
@dy will you have the time to look at these Mac issues brought up by @alexcjohnson over the next 48 hours?

**Comment by dy** [comment]
@etpinard working on that. That seems to be plotly issue, not gl-text. But Safari is broken as well.

**Comment by dy** [comment]
Ok, should be fixed!

**Comment by etpinard** [comment]
Thanks @dy !

Things have improved as far as I can tell on High Sierra on browserstack, the `gl2d_text_chart_styling` mock now looks like

in Chrome 67:

image

in Safari 11.1:

image

**but** something is still off in FF 61:

image

**Comment by dy** [comment]
@etpinard hmm, I am getting 
image
in FF 61 in OSX 10.10.5

**Comment by etpinard** [comment]
still getting:

image

in FF61 on High Sierra (on browserstack) and my Ubuntu 16.04 laptop.

**Comment by etpinard** [comment]
trace stack in ubuntu:

image

**Comment by dy** [comment]
@etpinard can you please try one more time? (blind debugging, but no choice)

**Comment by etpinard** [comment]
@dy for `gl2d_text_chart_arrays`, no errors but incorrect rendering:

image

**Comment by etpinard** [comment]
but it's weird, after refreshing (ctrl-r), I get:

image

**Comment by dy** [comment]
@etpinard is that permanent behavior?

**Comment by etpinard** [comment]
> is that permanent behavior?

what do mean by permanent? 

First render is a given tabs is incorrect, hitting refresh once or mulitple times give the correct view

**Comment by etpinard** [comment]
Hopefully @alexcjohnson  will have time to try this out in the next 15 hours or so. If not, I'm tempted to just merge this thing, flag it as "beta" in the release notes and have users try it out so that we can accumulate reports. Thoughts?

N.B. I'm about to go offline until about 9am tomorrow EDT (no internet at home after moving day).

**Comment by alexcjohnson** [comment]
Definite improvement, though there are still some oddities - I'm seeing similar effects to what @etpinard reported. I'd be OK with merging this as is and considering these issues bugs. Fairly urgent bugs though!

On Mac OS 10.13.5, retina display. For reference here's the SVG `text_chart_arrays` mock, which AFAICT looks identical on Chrome, FF, and Safari:
chrome svg
But the gl2d version looks bold and blurry (perhaps this is retina-only, related to #2756?), and there are first-load issues in FF (missing some items entirely) and Safari (wrong font).
Here's Chrome:
chrome gl
FF the first render in a given tab is missing some items:
ff gl first
But the second gets them all:
ff gl second
And Safari on first render has the wrong font for "Text G":
safari gl
but it's fixed on second render:
safari gl second

**Comment by etpinard** [comment]
Followup issue for @dy -> https://github.com/plotly/plotly.js/issues/2780

**Comment by etpinard** [comment]
Merging as "beta".

Please comment on https://github.com/plotly/plotly.js/issues/2780 if you're seeing odd behavior.

**Comment by etpinard** [comment]
Some benchmarks using:

```js
var n = /**/;
var letters = 'abcdefghijklmnopqrstuvwxyz';
var text = [];
var x = [];
var y = [];

for(var i = 0; i < n; i++) {
  x.push(Math.random())
  y.push(Math.random())
  text.push(letters[i % letters.length])
}

console.time('tx')
Plotly.newPlot(gd, [{
  type: 'scattergl', // or 'scatter'
  mode: 'text',
  x: x,
  y: y,
  text: text,
}])
console.timeEnd('tx')
```


| n | scattergl [in ms] | scatter [in ms] |
| -- | ----------- | --------- |
| 100 | 150 | 80 |
| 1e3 | 170 | 250 |
| 1e4 | 220 | 2000 |
| 1e5 | 500 | --- |
| 1e6 | 1500-2000 | --- |

But note that scattergl does not cluster text node, so panning and selection at when n > 1e5 is slow.

**Pull request merged**